nrf52840dongle build failure in zephyr

The `nRF52840dongle_nrf52840` board fails to build in the current 3.3 main of zephyr with the following error:

```
 file included from /tmp/aa/zephyr/include/zephyr/sys/util_macro.h:34,
                 from /tmp/aa/zephyr/include/zephyr/sys/util.h:17,
                 from /tmp/aa/zephyr/include/zephyr/devicetree.h:21,
                 from /tmp/aa/zephyr/include/zephyr/device.h:12,
                 from /tmp/aa/zephyr/include/zephyr/drivers/uart.h:26,
                 from /tmp/aa/bootloader/mcuboot/boot/zephyr/serial_adapter.c:18:
/tmp/aa/bootloader/mcuboot/boot/zephyr/serial_adapter.c: In function 'boot_uart_fifo_init':
/tmp/aa/zephyr/include/zephyr/sys/util.h:88:55: error: size of unnamed array is negative
   88 | #define ZERO_OR_COMPILE_ERROR(cond) ((int) sizeof(char[1 - 2 * !(cond)]) - 1)
      |                                                       ^
/tmp/aa/zephyr/include/zephyr/sys/util_internal.h:72:26: note: in definition of macro '__DEBRACKET'
   72 | #define __DEBRACKET(...) __VA_ARGS__
      |                          ^~~~~~~~~~~
/tmp/aa/zephyr/include/zephyr/sys/util_internal.h:64:9: note: in expansion of macro '__GET_ARG2_DEBRACKET'
   64 |         __GET_ARG2_DEBRACKET(one_or_two_args _if_code, _else_code)
      |         ^~~~~~~~~~~~~~~~~~~~
/tmp/aa/zephyr/include/zephyr/sys/util_internal.h:59:9: note: in expansion of macro '__COND_CODE'
   59 |         __COND_CODE(_XXXX##_flag, _if_1_code, _else_code)
      |         ^~~~~~~~~~~
/tmp/aa/zephyr/include/zephyr/sys/util_macro.h:177:9: note: in expansion of macro 'Z_COND_CODE_1'
  177 |         Z_COND_CODE_1(_flag, _if_1_code, _else_code)
      |         ^~~~~~~~~~~~~
/tmp/aa/zephyr/include/zephyr/device.h:277:9: note: in expansion of macro 'COND_CODE_1'
  277 |         COND_CODE_1(DT_HAS_COMPAT_STATUS_OKAY(compat),                         \
      |         ^~~~~~~~~~~
/tmp/aa/zephyr/include/zephyr/device.h:279:22: note: in expansion of macro 'ZERO_OR_COMPILE_ERROR'
  279 |                     (ZERO_OR_COMPILE_ERROR(0)))
      |                      ^~~~~~~~~~~~~~~~~~~~~
/tmp/aa/bootloader/mcuboot/boot/zephyr/serial_adapter.c:197:20: note: in expansion of macro 'DEVICE_DT_GET_ONE'
  197 |         uart_dev = DEVICE_DT_GET_ONE(zephyr_cdc_acm_uart);
      |                    ^~~~~~~~~~~~~~~~~
[75/300] Building C object zephyr/CMakeFiles/zephyr.dir/subsys/usb/device/usb_device.c.obj
ninja: build stopped: subcommand failed.
```

--- a/boards/arm/nrf52840dongle_nrf52840/Kconfig
+++ b/boards/arm/nrf52840dongle_nrf52840/Kconfig
@@ -1,6 +1,6 @@
 # nRF52840 Dongle NRF52840 board configuration
 
-# Copyright (c) 2018 Nordic Semiconductor ASA
+# Copyright (c) 2018-2023 Nordic Semiconductor ASA
 # SPDX-License-Identifier: Apache-2.0
 
 if BOARD_NRF52840DONGLE_NRF52840
@@ -21,5 +21,9 @@ config BOARD_HAS_NRF5_BOOTLOADER
 	help
 	  If selected, applications are linked so that they can be loaded by Nordic
 	  nRF5 bootloader.
+
+config BOARD_SERIAL_BACKEND_CDC_ACM
+	bool "USB CDC"
+	default y
 
 endif # BOARD_NRF52840DONGLE_NRF52840

--- a/boards/arm/nrf52840dongle_nrf52840/Kconfig.defconfig
+++ b/boards/arm/nrf52840dongle_nrf52840/Kconfig.defconfig
@@ -1,6 +1,6 @@
 # nRF52840 Dongle NRF52840 board configuration
 #
-# Copyright (c) 2018 Nordic Semiconductor ASA
+# Copyright (c) 2018-2023 Nordic Semiconductor ASA
 #
 # SPDX-License-Identifier: Apache-2.0
 
@@ -24,6 +24,44 @@ config FLASH_LOAD_OFFSET
 	default 0x1000
 	depends on BOARD_HAS_NRF5_BOOTLOADER && !USE_DT_CODE_PARTITION
 
+if BOARD_SERIAL_BACKEND_CDC_ACM
+
+config USB_DEVICE_STACK
+	default y
+
+config USB_CDC_ACM
+	default y
+
+config UART_CONSOLE
+	default CONSOLE
+
+config USB_DEVICE_INITIALIZE_AT_BOOT
+	default y if !MCUBOOT
+
+config SHELL_BACKEND_SERIAL_CHECK_DTR
+	default SHELL
+
+config USB_DEVICE_REMOTE_WAKEUP
+	default n
+
+if LOG
+
+# Logger cannot use itself to log
+choice USB_CDC_ACM_LOG_LEVEL_CHOICE
+	default USB_CDC_ACM_LOG_LEVEL_OFF
+endchoice
+
+# Set USB log level to error only
+choice USB_DEVICE_LOG_LEVEL_CHOICE
+	default USB_DEVICE_LOG_LEVEL_ERR
+endchoice
+
+# Wait 4000ms at startup for logging
+config LOG_PROCESS_THREAD_STARTUP_DELAY_MS
+	default 4000
+
+endif # LOG
+
 if USB_DEVICE_STACK
 
 # Enable UART driver, needed for CDC ACM
@@ -31,6 +69,8 @@ config SERIAL
 	default y
 
 endif # USB_DEVICE_STACK
+
+endif # BOARD_SERIAL_BACKEND_CDC_ACM
 
 config BT_CTLR
 	default BT

--- a/boards/arm/nrf52840dongle_nrf52840/nrf52840dongle_nrf52840.dts
+++ b/boards/arm/nrf52840dongle_nrf52840/nrf52840dongle_nrf52840.dts
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2018 Nordic Semiconductor ASA
+ * Copyright (c) 2018-2023 Nordic Semiconductor ASA
  * Copyright (c) 2017 Linaro Limited
  *
  * SPDX-License-Identifier: Apache-2.0
@@ -14,11 +14,11 @@
 	compatible = "nordic,nrf52840-dongle-nrf52840";
 
 	chosen {
-		zephyr,console = &uart0;
-		zephyr,shell-uart = &uart0;
-		zephyr,uart-mcumgr = &uart0;
-		zephyr,bt-mon-uart = &uart0;
-		zephyr,bt-c2h-uart = &uart0;
+		zephyr,console = &cdc_acm_uart;
+		zephyr,shell-uart = &cdc_acm_uart;
+		zephyr,uart-mcumgr = &cdc_acm_uart;
+		zephyr,bt-mon-uart = &cdc_acm_uart;
+		zephyr,bt-c2h-uart = &cdc_acm_uart;
 		zephyr,sram = &sram0;
 		zephyr,flash = &flash0;
 		zephyr,code-partition = &slot0_partition;
@@ -173,4 +173,8 @@
 zephyr_udc0: &usbd {
 	compatible = "nordic,nrf-usbd";
 	status = "okay";
+
+	cdc_acm_uart: cdc_acm_uart {
+		compatible = "zephyr,cdc-acm-uart";
+	};
 };

--- a/boards/arm/nrf52840dongle_nrf52840/nrf52840dongle_nrf52840_defconfig
+++ b/boards/arm/nrf52840dongle_nrf52840/nrf52840dongle_nrf52840_defconfig
@@ -18,3 +18,9 @@ CONFIG_GPIO_AS_PINRESET=y
 CONFIG_NFCT_PINS_AS_GPIOS=y
 
 CONFIG_PINCTRL=y
+
+# Board Kconfig.defconfig enables USB CDC ACM and should disable USB remote
+# wakeup by default. It needs to be disabled here, because the USB nrfx
+# driver always overwrites option from Kconfig mentioned above with the
+# imply from CONFIG_USB_NRFX.
+CONFIG_USB_DEVICE_REMOTE_WAKEUP=n

--- a/doc/releases/release-notes-3.3.rst
+++ b/doc/releases/release-notes-3.3.rst
@@ -347,6 +347,10 @@ Boards & SoC Support
     Note that MCUboot and MCUboot image updates from pre-Zephyr 3.3 might be
     incompatible with Zephyr 3.3 onwards and vice versa.
 
+  * The default console for the ``nrf52840dongle_nrf52840`` board has been
+    changed from physical UART (which is not connected to anything on the
+    board) to use USB CDC instead.
+
 * Made these changes in other boards:
 
 * Added support for these following shields:

